vncviewer/Viewport.cxx

```diff
@@ -1273,16 +1318,27 @@ int Viewport::handleSystemEvent(void *event, void *data)
               break;
           }
           break;
+
         case XI_TouchBegin:
-          buttonMask = 1;
+          if (gh->registerEvent(devev) == 1) {
+            // First touch. Give the user some time to complete
+	    // a gesture, then come back and process the events.
+            Fl::add_timeout(GH_TOUCHDELAY, touchTimeout, self);
```

Comment:
Can't we hide this inside `GestureHandler`? It seems odd that these details are forced upon `Viewport`.

Reply:
Maybe. There needs to be a way to make `Viewport` process the event queue immediately after timeout, without having to wait for another touch event to trigger the processing. Originally, I was playing around with `PropertyNotify` events, which are triggered whenever a window property changes. `GestureHandler` could modify a dummy window property to trigger a `PropertyNotify` event, and we could use this event to trigger processing of the queue (you can distinguish 'trigger' events based on the property which was changed). One potential benefit of doing this is that you get a current "X11" timestamp, which you can use to determine how old your tracked touches are. But this seems majorly convoluted. Open to ideas if you have any.

Reply:
Right now you have a "pull" approach where `Viewport` pulls events from `GestureHandler`. You could switch to a "push" approach instead, like most of the existing callbacks you are using. I.e. define an interface class (e.g. `GestureEventHandler`), have `Viewport` subclass it and register itself with the `GestureHandler` object. `GestureHandler` can then inject events whenever it damn well pleases.

Have a look at `rfb::Timer` and you'll see how this pattern works in practice.